### PR TITLE
allow input of gray color patches in arbitrary order

### PR DESCRIPTION
**Pull request by rabauke**
darktable-chart can read a list of source and target colors in LAB space from a csv file (or two images) to generate a style which maps the source colors to the target colors. This pull request addresses the following two issues.

* When processing the data in the current implementation of the function `process_data` it is assumed that the last `num_tonecurve` color values represent various shades of gray which are sorted with respect to the brightness of the source color, starting with the brightest shade. If the input data is not given in this particular sorted order a wrong tonecurve is calculated. The patch allows to give the grays in the input data in arbitrary order. It is sorted in `process_data` before further processing. This makes darktable-chart more robust and flexible with respect to input data in particular when given via a csv file.

* In the current implementation the tonecurve is calculated from a total of `num_tonecurve` data points given by the input data and two data points representing pure black in white. This means two of the gray input colors are actually discarded (the two brightest input colors are not used assuming the ordering mentioned above). The patched version of  `process_data` utilizes all `num_tonecurve` gray input colors **plus** pure black and white.

### CODE REVIEW

**Comment by LebedevRI** [comment]
> If the input data is not given in this particular sorted order a wrong tonecurve is calculated.

Yes but. What is the *actual* problem?
Are there color charts with such random order? From where does this unsorted problem comes from?

> This means two of the gray input colors are actually discarded (the two brightest input colors are not used assuming the ordering mentioned above)

I would *guess* it is a precausion against clipped colors that would throw off the fit, but then i don't know the real reasons..

**Comment by rabauke** [comment]
Dear Roman,

> Yes but. What is the actual problem?
> Are there color charts with such random order? From where does this unsorted problem comes from?

I came across this problem (and was not aware of this particular required ordering) when I used darktable-chart with csv files (as input) which have not been generated by darktable-chart. The option to feed directly LAB color values to darktable-chart via csv files allows to use this tool beyond creating styles out of pairs of images such as RAW+JPEG with in-camera processing.  E. g., this is extremely useful go generate color looks/color gradings. Furthermore, color checker with a more general ordering of gray patches exist. The [Video ColorChecker Passport by Xrite](http://www.xrite.com/service-support/product-support/calibration-solutions/colorchecker-passport-video) has two rows of grays, one with a gray scale ramp through the middle section of the tonal scale, one with shadows and highlighs.

> I would guess it is a precausion against clipped colors that would throw off the fit, but then i don't know the real reasons..

Many color checkers have only 6 gray patches. I do not think that it is a good idea to discard one third of them when building the tonecurve (without knowing that these will cause problems). If there is actually a good reason to do so, I would be curious to learn why.

Heiko

**Comment by houz** [comment]
I don't think that there is a real reason to discard some patches, it's most likely just a bug.

File: src/chart/main.c [hunk 1]
```diff
@@ -868,51 +868,77 @@ static char *encode_colorchecker(int num, const double *point, const double **ta
   return dt_exif_xmp_encode_internal((uint8_t *)&params, sizeof(params), NULL, FALSE);
 }
 
+static int compare_L_source(const void *x_, const void *y_) {
+  const double *x = x_;
+  const double *y = y_;
+  return (*x) < (*y) ? -1 : ((*x) > (*y) ? +1 : 0);
```
Comment (LebedevRI) [inline src/chart/main.c]:
```c++
static int compare_L_source(const void *x_, const void *y_) {
  const double x = *x_;
  const double y = *y_;
  return x < y ? -1 : (x > y ? +1 : 0);
}
```

File: src/chart/main.c [hunk 2]
```diff
@@ -868,51 +868,77 @@ static char *encode_colorchecker(int num, const double *point, const double **ta
...
+}
+
 static void process_data(dt_lut_t *self, double *target_L, double *target_a, double *target_b,
                          double *colorchecker_Lab, int N, int num_tonecurve, int sparsity)
 {
-  if (num_tonecurve < 2)
-    num_tonecurve = 2;
+  // sort all gray patches according to the L channel of the source colors
+  double *grays = malloc(sizeof(double) * 6 *num_tonecurve);
```
Comment (LebedevRI) [inline src/chart/main.c]:
```
double *grays = malloc(sizeof(double) * 6 * num_tonecurve);
```

File: src/chart/main.c (hunk 1 above)
Comment (rabauke) [reply src/chart/main.c]: This implementation of the comparison function is more easy to read, indeed.

File: src/chart/main.c (hunk 2 above)
Comment (rabauke) [reply src/chart/main.c]: These two lines of code are no longer required. Not removing them is a bug.

File: src/chart/main.c (hunk 1 above)
Comment (rabauke) [reply src/chart/main.c]: ... but this is not valid C. A type cast needs to be applied before dereferencing the pointers.

Comment (LebedevRI) [reply src/chart/main.c]: Ah well :)
So either
```c++
static int compare_L_source(const void *x_, const void *y_) {
  const double x = *(const double*)x_;
  const double y = *(const double*)y_;
  return x < y ? -1 : (x > y ? +1 : 0);
}
```

or
```c++
static int compare_L_source(const double *x_, const double *y_) {
  const double x = *x_;
  const double y = *y_;
  return x < y ? -1 : (x > y ? +1 : 0);
}
...
  qsort(grays, num_tonecurve, sizeof(double) * 6, (int (*)(const void *, const void *))compare_L_source);
```

Comment (rabauke) [reply src/chart/main.c]: Both variants will do the job. Personally I would prefer the 1st one. Function pointer casts are difficult to read.

File: src/chart/main.c (hunk 2 above)
Comment (rabauke) [reply src/chart/main.c]: I am unsure if I understand your comment. Please increase verbosity level.

Comment (LebedevRI) [reply src/chart/main.c]:
```diff
-+  double *grays = malloc(sizeof(double) * 6 *num_tonecurve);
++  double *grays = malloc(sizeof(double) * 6 * num_tonecurve);
```

In other words, what is `*num_tonecurve`?
Is `num_tonecurve` a pointer?
And then you have `6` and "dereferenced" pointer which wouldn't even compile.

TLDR: [.git/hooks/pre-commit](https://github.com/darktable-org/darktable/blob/master/CONTRIBUTING.md#coding-style)

Comment (rabauke) [reply src/chart/main.c]: I configured `git-clang-format` and revised the code. It should comply with the darktable coding style now.

**Comment by LebedevRI** [comment]
On Thu, Sep 14, 2017 at 11:34 AM, Heiko Bauke <notifications@github.com>
wrote:

> *@rabauke* commented on this pull request.
> ------------------------------
>
> In src/chart/main.c
> <https://github.com/darktable-org/darktable/pull/1532#discussion_r138829190>
> :
>
> >  static void process_data(dt_lut_t *self, double *target_L, double *target_a, double *target_b,
>                           double *colorchecker_Lab, int N, int num_tonecurve, int sparsity)
>  {
> -  if (num_tonecurve < 2)
> -    num_tonecurve = 2;
> +  // sort all gray patches according to the L channel of the source colors
> +  double *grays = malloc(sizeof(double) * 6 *num_tonecurve);
>
> I configured git-clang-format and revised the code. It should comply with
> the darktable coding style now.
>
Thanks, it does help. However when working on rather mis-formatted code,
the results may wary :/

Roman.
